curp/src/server/cmd_worker/conflict_checked_mpmc.rs

```diff
@@ -0,0 +1,387 @@
+#![allow(
+    clippy::wildcard_enum_match_arm,
+    clippy::match_wildcard_for_single_variants
+)] // wildcard actually is more clear in this module
+#![allow(clippy::integer_arithmetic)] // u64 is large enough
+
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use tracing::{debug, error};
+
+use super::{CEEvent, Task, TaskType};
+use crate::cmd::{Command, ProposeId};
+
+/// Vertex
+struct Vertex<C> {
+    /// Successor cmds that arrive later with keys that conflict this cmd
+    successors: HashSet<u64>,
+    /// Number of predecessor cmds that arrive earlier with keys that conflict this cmd
+    predecessor_cnt: u64,
+    /// Vertex inner
+    inner: VertexInner<C>,
+}
+
+impl<C: Command> Vertex<C> {
+    /// Whether two vertex conflict each other
+    fn is_conflict(&self, other: &Vertex<C>) -> bool {
+        #[allow(clippy::pattern_type_mismatch)]
+        // it seems it's impossible to get away with this lint
+        match (&self.inner, &other.inner) {
+            (VertexInner::Cmd { cmd: cmd1, .. }, VertexInner::Cmd { cmd: cmd2, .. }) => {
+                cmd1.is_conflict(cmd2.as_ref())
+            }
+            _ => true,
+        }
+    }
+}
+
+/// Vertex inner
+enum VertexInner<C> {
+    /// A cmd vertex
+    Cmd {
+        /// Cmd
+        cmd: Arc<C>,
+        /// Execution state
+        exe_st: ExeState,
+        /// After sync state
+        as_st: AsState,
+    },
+    /// A reset vertex
+    Reset(ResetState),
+}
+
+/// Execute state of a cmd
+#[derive(Debug, Clone, Copy)]
+enum ExeState {
+    /// Is ready to execute
+    ExecuteReady,
+    /// Executing
+    Executing,
+    /// Has been executed, and the result
+    Executed(bool),
+}
+
+/// After sync state of a cmd
+#[derive(Debug, Clone, Copy)]
+enum AsState {
+    /// Not Synced yet
+    NotSynced,
+    /// Is ready to do after sync
+    AfterSyncReady(usize),
+    /// Is doing after syncing
+    AfterSyncing,
+    /// Has been after synced
+    AfterSynced,
+}
+
+/// Reset state
+#[derive(Debug, Clone, Copy)]
+enum ResetState {
+    /// Reset ready
+    ResetReady,
+    /// Resetting
+    Resetting,
+    /// Completed
+    Completed,
+}
+
+/// The filter will block any msg if its predecessors(msgs that arrive earlier and conflict with it) haven't finished process
+/// Internally it maintains a dependency graph of conflicting cmds
+struct Filter<C> {
+    /// Index from `ProposeId` to `vertex`
+    cmd_id: HashMap<ProposeId, u64>,
```

Comment:
`cmd_id` implicitly indicates that it's a number. How about `cmd_to_vertex`?

Reply:
Fixed